modules/brc20_swap_index/lib/brc20_swap/indexer/module_commit_add_liq.go

```diff
@@ -0,0 +1,199 @@
+package indexer
+
+import (
+	"errors"
+	"log"
+
+	"github.com/unisat-wallet/libbrc20-indexer/conf"
+	"github.com/unisat-wallet/libbrc20-indexer/constant"
+	"github.com/unisat-wallet/libbrc20-indexer/decimal"
+	"github.com/unisat-wallet/libbrc20-indexer/model"
+	"github.com/unisat-wallet/libbrc20-indexer/utils"
+)
+
+func (g *BRC20ModuleIndexer) ProcessCommitFunctionAddLiquidity(moduleInfo *model.BRC20ModuleSwapInfo, f *model.SwapFunctionData) (err error) {
+	token0, token1 := f.Params[0], f.Params[1]
+	if g.BestHeight < conf.ENABLE_SWAP_WITHDRAW_HEIGHT {
+		token0, token1, err = utils.DecodeTokensFromSwapPair(f.Params[0])
+		if err != nil {
+			return errors.New("func: addLiq poolPair invalid")
+		}
+	}
+
+	poolPair := GetLowerInnerPairNameByToken(token0, token1)
+	pool, ok := moduleInfo.SwapPoolTotalBalanceDataMap[poolPair]
+	if !ok {
+		return errors.New("addLiq: pool invalid")
+	}
+
+	usersLpBalanceInPool, ok := moduleInfo.LPTokenUsersBalanceMap[poolPair]
+	if !ok {
+		return errors.New("addLiq: users invalid")
+	}
+
+	// log.Printf("[%s] pool before addliq [%s] %s: %s, %s: %s, lp: %s", moduleInfo.ID, poolPair, pool.Tick[0], pool.TickBalance[0], pool.Tick[1], pool.TickBalance[1], pool.LpBalance)
+	log.Printf("pool addliq params: %v", f.Params)
+
+	offset := 0
+	if g.BestHeight >= conf.ENABLE_SWAP_WITHDRAW_HEIGHT {
+		offset = 1
+	}
+	token0AmtStr := f.Params[1+offset]
+	token1AmtStr := f.Params[2+offset]
+	tokenLpAmtStr := f.Params[3+offset]
+
+	token0Amt, _ := g.CheckTickVerify(token0, token0AmtStr)
+	token1Amt, _ := g.CheckTickVerify(token1, token1AmtStr)
+	tokenLpAmt, _ := decimal.NewDecimalFromString(tokenLpAmtStr, 18)
+
+	// LP Balance Slippage Check
+	slippageAmtStr := f.Params[4+offset]
+	slippageAmt, _ := decimal.NewDecimalFromString(slippageAmtStr, 3)
+
+	var token0Idx, token1Idx int
+	if token0 == pool.Tick[0] {
+		token0Idx = 0
+		token1Idx = 1
+	} else {
+		token0Idx = 1
+		token1Idx = 0
+	}
+
+	var first bool = false
+	var lpForPool, lpForUser *decimal.Decimal
+	if pool.TickBalance[0].Sign() == 0 && pool.TickBalance[1].Sign() == 0 {
+		first = true
+		lpForPool = token0Amt.Mul(token1Amt).Sqrt()
+		if lpForPool.Cmp(decimal.NewDecimal(1000, 18)) < 0 {
+			return errors.New("addLiq: lp less than 1000")
+		}
+		lpForUser = lpForPool.Sub(decimal.NewDecimal(1000, 18))
+
+	} else {
+		// Issuing additional LP, as a way of collecting service fees.
+		feeRateSwapAmt, ok := CheckAmountVerify(moduleInfo.FeeRateSwap, 3)
+		if !ok {
+			log.Printf("pool addliq FeeRateSwap invalid: %s", moduleInfo.FeeRateSwap)
+			return errors.New("addLiq: feerate swap invalid")
+		}
+		if feeRateSwapAmt.Sign() > 0 {
+			// lp = (poolLp * (rootK - rootKLast)) / (rootK * 5 + rootKLast)
+			rootK := pool.TickBalance[token0Idx].Mul(pool.TickBalance[token1Idx]).Sqrt()
```

Comment:
Not sure if I'm missing anything but rootK here seems like it will be the same as lastRootK, because the tick balances didn't change in-between, making (rootK - rootKLast) always 0.